File: OpenStreetMap/OSMiOS/OSMiOS/FpsLabel.m
//
//  FpsLabel.m
//  Go Map!!
//
//  Created by Bryce on 6/15/14.
//  Copyright (c) 2014 Bryce. All rights reserved.
//

#import "FpsLabel.h"


const int FRAME_COUNT = 61;


@implementation FpsLabel
{
	int					_historyPos;
	CFTimeInterval		_history[ FRAME_COUNT ];	// average last 60 frames
	dispatch_source_t	_timer;
	CADisplayLink	*	_displayLink;
}

- (void)awakeFromNib
{
	[super awakeFromNib];

#if 0 && defined(DEBUG)
	_displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(displayLink)];
	[_displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];

	dispatch_queue_t queue = dispatch_get_main_queue();
	_timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
	if ( _timer ) {
		dispatch_source_set_timer(_timer, DISPATCH_TIME_NOW, NSEC_PER_SEC/2, NSEC_PER_SEC/5);
		dispatch_source_set_event_handler(_timer, ^{
			[self updateText];
		} );
		dispatch_resume(_timer);
	}
#else
	self.text = nil;
	[self removeFromSuperview];
#endif
}

- (void)dealloc
{
	dispatch_source_cancel( _timer );
}

- (void)displayLink
{
	[self frameUpdated];
}

- (void)updateText
{
	CFTimeInterval now = CACurrentMediaTime();

	// scan backward to see how many frames were drawn in the last second
	int frameCount = 0;
	int pos = _historyPos;
	CFTimeInterval prev = 0.0;
	do {
		if ( --pos < 0 )
			pos = FRAME_COUNT - 1;
		prev = _history[pos];
		++frameCount;
		if ( now - prev >= 1.0 )
			break;
	} while ( pos != _historyPos );

	CFTimeInterval average = frameCount / (now - prev);
	if ( average >= 10.0 )
		self.text = [NSString stringWithFormat:@"%.1f FPS", average];
	else
		self.text = [NSString stringWithFormat:@"%.2f FPS", average];
}

- (void)frameUpdated
{
#if defined(DEBUG)
	// add to history
	CFTimeInterval now = CACurrentMediaTime();
	_history[_historyPos++] = now;
	if ( _historyPos >= FRAME_COUNT )
		_historyPos = 0;

	[self updateText];
#endif
}

@end

File: src/Shared/FpsLabel.swift
//
//  FpsLabel.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 6/15/14.
//  Copyright (c) 2014 Bryce Cogswell. All rights reserved.
//

import UIKit

fileprivate let ENABLE_FPS = 0

fileprivate let FRAME_COUNT = 2 * 60

class FpsLabel: UILabel {
    private var historyPos = 0
    private var frameTimestamp = [CFTimeInterval](repeating: 0, count: FRAME_COUNT) // average last 60 frames
    
    private var timer: DispatchSourceTimer?

    private var _showFPS = false
    var showFPS: Bool {
        get {
            _showFPS
        }
        set(showFPS) {
            if showFPS != _showFPS {
                _showFPS = showFPS

                if showFPS {
                    isHidden = false
                    let displayLink = DisplayLink.shared()
                    displayLink.addName("FpsLabel", block: {
                        self.frameUpdated()
                    })

                    // create a timer to update the text twice a second
                    timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
					timer?.schedule(deadline: .now(), repeating: .milliseconds(500))
                    timer?.setEventHandler(handler: { [weak self] in
                        self?.updateText()
                    })
                    timer?.activate()
                    layer.backgroundColor = UIColor(white: 1.0, alpha: 0.6).cgColor
                } else {
                    text = nil
                    isHidden = true
                    let displayLink = DisplayLink.shared()
                    displayLink.removeName("FpsLabel")
                    timer?.cancel()
                }
            }
        }
    }

    override func awakeFromNib() {
        super.awakeFromNib()
        showFPS = false
        isHidden = true
    }

    deinit {
        showFPS = false
    }

    func updateText() {
        // scan backward to see how many frames were drawn in the last second
        let seconds = Double(FRAME_COUNT / 60)
        var frameCount: Double = 0
        var pos = (historyPos + FRAME_COUNT - 1) % FRAME_COUNT
        let last = frameTimestamp[pos]
        var prev: CFTimeInterval = 0.0
        repeat {
            pos -= 1
            if pos < 0 {
                pos = FRAME_COUNT - 1
            }
            prev = frameTimestamp[pos]
            frameCount += 1
            if last - prev >= seconds {
                break
            }
        } while pos != historyPos

        let average = CFTimeInterval(frameCount / (last - prev))
        if average >= 10.0 {
            text = String(format: "%.1f FPS", average)
        } else {
            text = String(format: "%.2f FPS", average)
        }
    }

    func frameUpdated() {
        // add to history
        let now = CACurrentMediaTime()
        frameTimestamp[historyPos] = now
        historyPos += 1
        if historyPos >= FRAME_COUNT {
            historyPos = 0
        }
    }
}
